File: __tests__/L3_CustodyParity.test.ts
/**
 * L3 Phase 1a — Parity reference vectors (JS side).
 *
 * This is the JS half of the de-risking harness. It runs the REAL production
 * primitives (the exact functions today's vaults are written with) and:
 *   1. asserts internal self-consistency (encrypt -> decrypt round-trips, MAC verifies);
 *   2. emits a vectors file the native/JCA harness (L3ParityHarness.java) re-derives
 *      and compares against, byte for byte.
 *
 * No production code is touched. If the native side diverges by a single byte, that is
 * a real finding: the file-key path cannot move to native custody without a wire-format
 * change (forbidden), so L3-file-key stays honestly Partial.
 *
 * Primitives exercised (all master-key consumers from the Phase 0 map):
 *   - AES-256-CBC/PKCS7  (aesCbcEncryptRaw/Raw)            -> JCA AES/CBC/PKCS5Padding
 *   - HMAC-SHA256        (SecureCryptoService.computeMac)  -> JCA HmacSHA256
 *   - HKDF-SHA256        (hkdfSha256, label filevault-mac-v1) -> RFC5869 over Mac
 *   - Encrypt-then-MAC file-key wrap + master-blob wrap    -> the real "pre-rebuild blob"
 */

const secureStore = require('../__mocks__/expo-secure-store');

import * as fs from 'fs';
import { SecureCryptoService } from '../src/services/CryptoService';
import { aesCbcEncryptRaw, aesCbcDecryptRaw } from '../src/services/AesCbcHmac';
import { hkdfSha256 } from '../src/services/FastPBKDF2';

const OUT =
  process.env.L3_VECTORS_OUT ||
  '/tmp/claude-1000/-mnt-c-Users-linux-Desktop-Coding-Projekte-Obscuraa-FileVault/53f1acbf-4355-4182-bf8d-536d14462dc5/scratchpad/l3_vectors.json';

// Fixed, deterministic test material so the vectors are reproducible and can also be
// hard-coded into the on-device androidTest later.
const KEY = '00112233445566778899aabbccddeeff102132435465768798a9bacbdcedfe0f'; // 32 bytes
const IV  = '0f1e2d3c4b5a69788796a5b4c3d2e1f0';                                 // 16 bytes
const utf8B64 = (s: string) => Buffer.from(s, 'utf8').toString('base64');

type AesVec = { name: string; keyHex: string; ivHex: string; plainB64: string; ctHex: string; macHex: string; macKeyHex: string };
type HkdfVec = { ikmHex: string; info: string; len: number; outHex: string };
type HmacVec = { name: string; keyHex: string; msgB64: string; outHex: string };
type BlobVec = { name: string; keyHex: string; ivHex: string; plainB64: string; ctHex: string; macHex: string };

const aes: AesVec[] = [];
const hkdf: HkdfVec[] = [];
const hmac: HmacVec[] = [];
const blobs: BlobVec[] = [];

describe('L3 Phase 1a: JS parity reference', () => {
  test('AES-256-CBC/PKCS7 + EtM-MAC at several lengths incl. exact block boundary', async () => {
    const macKeyHex = await SecureCryptoService.deriveMacKey(KEY);
    // 16 = exact block boundary: PKCS7 MUST add a full 16-byte padding block (-> 32 ct bytes).
    // 32, 64 = further exact multiples; 64 is the real file-key plaintext length.
    for (const len of [1, 15, 16, 17, 31, 32, 63, 64]) {
      const plain = Array.from({ length: len }, (_, i) => String.fromCharCode(33 + (i % 90))).join('');
      const ctHex = await aesCbcEncryptRaw(plain, KEY, IV);
      const macHex = await SecureCryptoService.computeMac(macKeyHex, IV + ctHex);
      // self-consistency: the production decrypt path must recover the same bytes
      expect(await aesCbcDecryptRaw(ctHex, KEY, IV)).toBe(plain);
      // block-boundary invariant: ct length (bytes) = ceil((len+1)/16)*16
      const expectedCtBytes = (Math.floor(len / 16) + 1) * 16;
      expect(ctHex.length).toBe(expectedCtBytes * 2);
      aes.push({ name: `len${len}`, keyHex: KEY, ivHex: IV, plainB64: utf8B64(plain), ctHex, macHex, macKeyHex });
    }
  }, 30000);

  test('HKDF-SHA256(filevault-mac-v1) + HMAC-SHA256 vectors', async () => {
    for (const ikm of [KEY, 'a'.repeat(64), 'ff'.repeat(32)]) {
      hkdf.push({ ikmHex: ikm, info: 'filevault-mac-v1', len: 32, outHex: hkdfSha256(ikm, 'filevault-mac-v1', 32) });
    }
    const mk = await SecureCryptoService.deriveMacKey(KEY);
    for (const msg of ['', 'abc', IV + 'deadbeef', 'x'.repeat(200)]) {
      hmac.push({ name: `msg${msg.length}`, keyHex: mk, msgB64: utf8B64(msg), outHex: await SecureCryptoService.computeMac(mk, msg) });
    }
    expect(hkdf[0].outHex).toHaveLength(64);
  });

  test('real pre-rebuild blobs: file-key wrap + master-blob (EtM)', async () => {
    // File-key wrap: exactly what FileManager stores per file (encryptFileKeyWith).
    const fileKey = 'deadbeefcafef00d' + '1122334455667788'.repeat(3); // 64 hex chars
    const wrap = await SecureCryptoService.encryptFileKeyWith(fileKey, KEY);
    expect(await SecureCryptoService.decryptFileKeyWith(wrap, KEY)).toBe(fileKey);
    blobs.push({ name: 'filekey-wrap', keyHex: KEY, ivHex: wrap.iv, plainB64: utf8B64(fileKey), ctHex: wrap.encryptedKey, macHex: wrap.mac });

    // Master-blob: same construction wrapAndStoreMasterKey uses (master wrapped under KEK).
    const kek = '9988776655443322110ffeeddccbbaa00112233445566778899aabbccddeeff0';
    const masterPlain = 'cafebabe00ff'.padEnd(64, '0');
    const mwrap = await SecureCryptoService.encryptFileKeyWith(masterPlain, kek);
    expect(await SecureCryptoService.decryptFileKeyWith(mwrap, kek)).toBe(masterPlain);
    blobs.push({ name: 'master-blob', keyHex: kek, ivHex: mwrap.iv, plainB64: utf8B64(masterPlain), ctHex: mwrap.encryptedKey, macHex: mwrap.mac });
  }, 30000);

  afterAll(() => {
    fs.writeFileSync(OUT, JSON.stringify({ aes, hkdf, hmac, blobs }, null, 2));
    // Flat, pipe-delimited mirror so the plain-JDK harness needs no JSON lib.
    // All fields are hex/base64/label -> no '|' collisions.
    const lines: string[] = [];
    for (const a of aes) lines.push(`AES|${a.name}|${a.keyHex}|${a.ivHex}|${a.plainB64}|${a.ctHex}|${a.macHex}|${a.macKeyHex}`);
    for (const h of hkdf) lines.push(`HKDF|${h.ikmHex}|${h.info}|${h.len}|${h.outHex}`);
    for (const m of hmac) lines.push(`HMAC|${m.name}|${m.keyHex}|${m.msgB64}|${m.outHex}`);
    for (const b of blobs) lines.push(`BLOB|${b.name}|${b.keyHex}|${b.ivHex}|${b.plainB64}|${b.ctHex}|${b.macHex}`);
    fs.writeFileSync(OUT.replace(/\.json$/, '.txt'), lines.join('\n') + '\n');
    // eslint-disable-next-line no-console
    console.log(`[L3-parity] wrote ${aes.length + hkdf.length + hmac.length + blobs.length} vectors -> ${OUT}`);
  });
});

File: android/app/src/androidTest/java/com/filevault/app/L3CustodyParityTest.kt
package com.filevault.app

import com.filevault.app.modules.NativeKeyCustodyCrypto
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test
import org.junit.runner.RunWith
import androidx.test.ext.junit.runners.AndroidJUnit4

/**
 * L3 Phase 1a — on-device parity Known-Answer Test for NativeKeyCustodyCrypto.
 *
 * Pins the native JCA path to the SAME golden vectors the JS production primitives emit
 * (__tests__/L3_CustodyParity.test.ts -> tools/parity/L3ParityHarness.java, host-proven 45/45).
 * Confirms on the real Android security provider that:
 *   1. AES-256-CBC/PKCS7 == CryptoJS PKCS7, incl. the exact block-boundary case (16 -> 32 bytes);
 *   2. HKDF-SHA256(filevault-mac-v1) + HMAC-SHA256 == deriveMacKey/computeMac;
 *   3. a pre-rebuild Encrypt-then-MAC blob (64-byte file-key plaintext -> 80-byte ct)
 *      verifies + decrypts + re-encrypts byte-identical through the native path.
 *
 * If any vector diverges, the file-key path cannot move to native custody without changing
 * the wire format (forbidden) -> L3-file-key stays Partial. No production code is wired to
 * NativeKeyCustodyCrypto yet; this test is its only caller.
 */
@RunWith(AndroidJUnit4::class)
class L3CustodyParityTest {

    private val KEY = NativeKeyCustodyCrypto.fromHex(
        "00112233445566778899aabbccddeeff102132435465768798a9bacbdcedfe0f")
    private val IV = NativeKeyCustodyCrypto.fromHex("0f1e2d3c4b5a69788796a5b4c3d2e1f0")

    // plain[i] = char(33 + i % 90), matching the JS reference generator (ASCII).
    private fun plain(len: Int): ByteArray =
        (0 until len).map { (33 + it % 90).toChar() }.joinToString("").toByteArray(Charsets.UTF_8)

    private val MACKEY_HEX = "9b10c116efd8dffd4b0e30709d4af53725daf41d067f87250719a7bf747197ce"
    private val LEN16_CT = "022ae7d0cd5faf24b4e0eb876a4ec72b23aa5c9dd1ecddd170c986dbbb72f76b"
    private val LEN64_CT = "022ae7d0cd5faf24b4e0eb876a4ec72b8d16c3b2605112040878bd74eced6027" +
        "ea64182be6550e67b30c91df8cfdaefdcba95faf5c70237fb17e671ad30b4e81b9556c1c7231f21c47ef789bc3c35c94"
    private val LEN64_MAC = "c4a2ce82f195e31c83540fec63dbbf020de4ffc50a09bee29b2be128b5c2c0d7"

    @Test
    fun aesCbcMatchesCryptoJsAtBlockBoundary() {
        // 16-byte input -> PKCS7 adds a full padding block -> 32-byte ciphertext.
        val ct16 = NativeKeyCustodyCrypto.aesCbcEncrypt(KEY, IV, plain(16))
        assertEquals(32, ct16.size)
        assertEquals(LEN16_CT, NativeKeyCustodyCrypto.toHex(ct16))
        assertEquals(
            String(plain(16), Charsets.UTF_8),
            String(NativeKeyCustodyCrypto.aesCbcDecrypt(KEY, IV, ct16), Charsets.UTF_8))

        // 64-byte input (file-key size) -> 80-byte ciphertext.
        val ct64 = NativeKeyCustodyCrypto.aesCbcEncrypt(KEY, IV, plain(64))
        assertEquals(80, ct64.size)
        assertEquals(LEN64_CT, NativeKeyCustodyCrypto.toHex(ct64))
    }

    @Test
    fun hkdfAndHmacMatchJs() {
        assertEquals(MACKEY_HEX, NativeKeyCustodyCrypto.toHex(NativeKeyCustodyCrypto.deriveMacKey(KEY)))
        assertEquals(
            "7eb3dc397bbca0b17cae61282b8e4e1ffb0d242a68a265d573307f0a6bd9daae",
            NativeKeyCustodyCrypto.toHex(
                NativeKeyCustodyCrypto.hmacSha256(
                    NativeKeyCustodyCrypto.fromHex(MACKEY_HEX), "abc".toByteArray(Charsets.UTF_8))))
    }

    @Test
    fun etmWrapUnwrapBlobByteIdentical() {
        // Wrap == the stored Encrypt-then-MAC blob.
        val w = NativeKeyCustodyCrypto.wrapEtm(KEY, IV, plain(64))
        assertEquals(LEN64_CT, w.ctHex)
        assertEquals(LEN64_MAC, w.macHex)

        // Pre-rebuild blob decrypts byte-identical through the native path.
        val dec = NativeKeyCustodyCrypto.unwrapEtm(KEY, NativeKeyCustodyCrypto.toHex(IV), LEN64_CT, LEN64_MAC)
        assertEquals(String(plain(64), Charsets.UTF_8), String(dec, Charsets.UTF_8))

        // Tampered tag -> reject, no decrypt.
        val badMac = LEN64_MAC.dropLast(2) + "00"
        assertThrows(SecurityException::class.java) {
            NativeKeyCustodyCrypto.unwrapEtm(KEY, NativeKeyCustodyCrypto.toHex(IV), LEN64_CT, badMac)
        }
    }
}
